File: baekjoon/15649_N과_M(1).py
# 백트래킹
# dfs를 이용하여 순열 print

import sys

n, m = map(int, sys.stdin.readline().split())
nums = [i+1 for i in range(n)]
visited = [False] * 8
result = [0] * 8

def dfs(lv):
    if lv == m:
        for k in range(m):
            print(result[k], end=' ')
        print('')
        return
    for i in range(n):
        if visited[i] == False:
            visited[i] = True
            result[lv] = nums[i]
            dfs(lv+1)
            visited[i] = False
        elif visited[i] == True:
            continue
            
dfs(0)

File: baekjoon/baekjoon/15649_N과_M(1).cpp
#include <iostream>
using namespace std;

int n, m;
bool visited[9] = {false, };
int nums[9] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
int result[9] = {0, };

void DFS(int lv) {
    if(lv == m) { // 재귀 종료조건
        for(int i=0; i<m; i++) {
            cout << result[i] << ' ';
        }
        cout << '\n';
        return;
    }

    for(int i=0; i<n; i++) {
        if(visited[i] == false) {
            visited[i] = true;
            result[lv] = nums[i];
            DFS(lv + 1);
            visited[i] = false;
        }
    }
}

int main() {
    cin >> n >> m;
    DFS(0);
}